### PR TITLE
[Bug Fixed] fix batch norm when `fix_gamma` is True

### PR DESCRIPTION
**Pull request by wkcn**
## Description ##
Fix the issue #16297 #18475
When `fix_gamma` is True, the batch norm will set `grad_req` of `gamma` to `null`.
CudnnBatchNorm will be invoked when `cudnn_off=False and axis=1`, and the gradient of `beta` will be accumulated by mistake.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Fix the bug of BatchNorm when `fix_gamma=True`.
- Add unittest for fixing gamma.

### CODE REVIEW

**Comment by wkcn** [comment]
Hi @sxjscience , could you please help take a review?

File: src/operator/nn/cudnn/cudnn_batch_norm-inl.h [hunk 1]
```diff
@@ -228,7 +228,7 @@ class CuDNNBatchNormOp {
         &a,
         &b,
         &a,
-        req[cudnnbatchnorm::kGamma] == kWriteTo ? &b: &b_add,
+        req[cudnnbatchnorm::kGamma] == kAddTo ? &b_add : &b,
```
Comment (sxjscience) [inline src/operator/nn/cudnn/cudnn_batch_norm-inl.h:231]: Here, we just test with `req[cudnnbatchnorm::kGamma] == kAddTo`. We won't be able to set the gradient accumulation appropriately if `req[cudnnbatchnorm::kGamma] == kWriteTo && req[cudnnbatchnorm::kBeta] == kAddTo`.

Comment (sxjscience) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:231]: @zhreshold I believe there are some `grad_req = add` issues in BN. Would you know if there are cases that `gamma.grad_req = write` and `beta.grad_req = add`? If not, we may just raise an error in the OP if it happens.

Comment (zhreshold) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:231]: I don't think there's any use case I can imagine, raising an exception is acceptable

Comment (wkcn) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:231]: @sxjscience @zhreshold Thanks for your review! I will add the grad_req check and an unittest for gradient accumulation.

Comment (wkcn) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:231]: `gamma.grad_req = kAddTo` is not supported in [Naive CPU Implementation](https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/batch_norm.cc#L304), [Naive GPU one](https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/batch_norm.cu#L383) and [MKLDNN one](https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h#L419).

Comment (sxjscience) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:231]: I believe it shows a larger problem of the BN layer...

Comment (wkcn) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:231]: The gradient of gamma on CPU is wrong when `grad_req` is True.
```python
import mxnet as mx
from mxnet.gluon import nn

N = 1
C = 3
H = W = 2
block = nn.BatchNorm() 
block.collect_params().initialize()
block.collect_params().setattr('grad_req', 'add')

for i in range(1):
    with mx.autograd.record():
        x = mx.nd.arange(N*C*H*W).reshape((N, C, H, W))
        y = block(x)
        loss = (y * y).sum() 
    loss.backward()
print(block.gamma.grad())
```

Comment (sxjscience) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:231]: I guess this is somehow out-of-the-scope of this PR. How about to create an issue and we can have another PR to fix the problem?

Comment (wkcn) [reply src/operator/nn/cudnn/cudnn_batch_norm-inl.h:231]: @sxjscience I will create an issue.

**Comment by wkcn** [comment]
Close it and focus on the PR #18500 , which contains this PR.
